frame/support/src/storage/mod.rs

```diff
@@ -48,46 +48,39 @@ pub mod unhashed;
 pub mod weak_bounded_vec;
 
 mod transaction_level_tracker {
-	use core::sync::atomic::{AtomicU32, Ordering};
+	use crate::sync::RuntimeCell;
 
 	type Layer = u32;
-	static NUM_LEVELS: AtomicU32 = AtomicU32::new(0);
+	static NUM_LEVELS: RuntimeCell<u32> = RuntimeCell::new(0);
```

Comment:
As long as we have the native runtime, this isn't safe. The native runtime can be executed from different threads.

Reply:
Really? Multiple threads operate on the same memory at the same time? Why would that be the case?

Reply:
Native runtime. This static is shared within the entire context of the binary.

Reply:
Sure but you still don't call into any runtime code concurrently with more than one thread? That would make no sense.

Reply:
Not the runtime itself, however you can call from different threads into the runtime. Imagine some thread importing a block and another validating an extrinsic. Both will call into the runtime

Reply:
I assume in the wasm case we spawn different instances for those calls which are then called in parallel.

To me it sounds like those calls need to be serialized to have the same behavior as with the wasm runtime.

Reply:
How do you want to serialize them? The point is, we don't use any global context and if we use it it needs to be thread local.

Reply:
Okay I get it now. The native runtime recycles the memory from other calls while the wasm gets fresh memory. Yeah this makes this wrong, unfortunately.